En este curso, se llevará a cabo un proyecto en el que se desempeñará el rol de científicos de datos. Se abordará el proceso inicial de familiarización con la información, así como el tratamiento y la limpieza de los datos utilizando Pandas para abordar los desafíos en cada una de las etapas.

El proyecto involucra una empresa ficticia que tiene la intención de utilizar productos destinados al mercado inmobiliario. Para este propósito, la empresa dispone de una base de datos con la cual trabajar, aunque se reconoce que esta base de datos puede presentar algunas dificultades en su estructura. El objetivo del proyecto es elaborar análisis e informes a partir de esta base de datos.

#**Reporte de Analisis I**

**Importando la Base de Datos**

In [3]:
import pandas as pd
pd.set_option("display.max_rows", 20)
pd.set_option("display.max_columns", 9)


Importando y montando mi unidad de drive.

In [4]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


Ruta completa al archivo en mi unidad de Google Drive.

SEP = separador.

le especifico el tipo de separador que usa el archivo(csv en este caso que separa los valores por comas) para una mejor visualizacion de los datos.

In [5]:
dataset_alquileres = pd.read_csv("/content/gdrive/MyDrive/Archivos-csv/Alquiler_Inmuebles/alquiler.csv", sep = ";")

#**Informaciones generales sobre la Base de Datos**

In [6]:
dataset_alquileres

Tipo    Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0           Habitación   San Borja        1         0       0    40  1700.0   
1                 Casa  Lurigancho        2         0       1   100  7000.0   
2      Local comercial         Ate        0         4       0   150  5200.0   
3         Departamento  Pachacámac        1         0       0    15   800.0   
4         Departamento         Ate        1         0       0    48   800.0   
...                ...         ...      ...       ...     ...   ...     ...   
32955       Habitación  Pachacámac        0         0       0    27   800.0   
32956     Departamento       Lince        3         1       2    78  1800.0   
32957     Departamento       Rímac        2         1       0    48  1400.0   
32958     Departamento       Rímac        2         0       0    70  3000.0   
32959  Local comercial  Pachacámac        0         0       0   250  6500.0   

       Mantenimiento  Impuesto  
0              500.0      60.0  
1                NaN       NaN  
2             4020.0    1111.0  
3              390.0      20.0  
4              230.0       NaN  
...              ...       ...  
32955          350.0      25.0  
32956          800.0      40.0  
32957          509.0      37.0  
32958          760.0       NaN  
32959         4206.0    1109.0  

[32960 rows x 9 columns]

In [7]:
dataset_alquileres.describe()

Cuartos      Vacantes        Suites          Area         Valor  \
count  32960.000000  32960.000000  32960.000000  32960.000000  3.294300e+04   
mean       1.771390      1.749059      0.665777    231.901547  1.295266e+04   
std        1.717869     20.380402      1.176525   1135.254152  6.675220e+05   
min        0.000000      0.000000      0.000000      0.000000  7.500000e+01   
25%        0.000000      0.000000      0.000000     55.000000  1.500000e+03   
50%        2.000000      1.000000      0.000000     87.000000  2.800000e+03   
75%        3.000000      2.000000      1.000000    170.000000  6.500000e+03   
max      100.000000   1966.000000     70.000000  90000.000000  1.200000e+08   

       Mantenimiento      Impuesto  
count   2.886700e+04  2.272300e+04  
mean    2.388062e+03  2.364410e+03  
std     3.918496e+04  1.795641e+05  
min     1.000000e+00  1.000000e+00  
25%     5.900000e+02  1.100000e+02  
50%     9.900000e+02  2.500000e+02  
75%     1.800000e+03  6.970000e+02  
max     6.552570e+06  2.705307e+07

In [8]:
type(dataset_alquileres)

pandas.core.frame.DataFrame

In [9]:
dataset_alquileres.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32960 entries, 0 to 32959
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Tipo           32960 non-null  object 
 1   Distrito       32960 non-null  object 
 2   Cuartos        32960 non-null  int64  
 3   Vacantes       32960 non-null  int64  
 4   Suites         32960 non-null  int64  
 5   Area           32960 non-null  int64  
 6   Valor          32943 non-null  float64
 7   Mantenimiento  28867 non-null  float64
 8   Impuesto       22723 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 2.3+ MB


In [10]:
dataset_alquileres.head(10)

Tipo     Distrito  Cuartos  Vacantes  Suites  Area    Valor  \
0          Habitación    San Borja        1         0       0    40   1700.0   
1                Casa   Lurigancho        2         0       1   100   7000.0   
2     Local comercial          Ate        0         4       0   150   5200.0   
3        Departamento   Pachacámac        1         0       0    15    800.0   
4        Departamento          Ate        1         0       0    48    800.0   
5        Departamento     Barranco        3         1       0    70   1200.0   
6        Departamento        Comas        2         0       0    50   1300.0   
7  Casa en condominio          Ate        5         4       5   750  22000.0   
8  Casa en condominio  El Agustino        2         2       0    65   1000.0   
9     Local comercial   Pachacámac        0         3       0   695  35000.0   

   Mantenimiento  Impuesto  
0          500.0      60.0  
1            NaN       NaN  
2         4020.0    1111.0  
3          390.0      20.0  
4          230.0       NaN  
5            NaN       NaN  
6          301.0      17.0  
7            NaN       NaN  
8            NaN       NaN  
9        19193.0    3030.0

In [11]:
dataset_alquileres.dtypes

Tipo              object
Distrito          object
Cuartos            int64
Vacantes           int64
Suites             int64
Area               int64
Valor            float64
Mantenimiento    float64
Impuesto         float64
dtype: object

pasando el formato de la informacion obtenida a df.

In [12]:
info_datos = pd.DataFrame(dataset_alquileres.dtypes)

In [13]:
info_datos

0
Tipo            object
Distrito        object
Cuartos          int64
Vacantes         int64
Suites           int64
Area             int64
Valor          float64
Mantenimiento  float64
Impuesto       float64

Actualizando la variable.

In [14]:
info_datos = pd.DataFrame(dataset_alquileres.dtypes, columns = ["Tipos de datos"])

In [15]:
info_datos

Tipos de datos
Tipo                  object
Distrito              object
Cuartos                int64
Vacantes               int64
Suites                 int64
Area                   int64
Valor                float64
Mantenimiento        float64
Impuesto             float64

In [16]:
info_datos.columns.name = "Variables"

In [17]:
info_datos

Variables     Tipos de datos
Tipo                  object
Distrito              object
Cuartos                int64
Vacantes               int64
Suites                 int64
Area                   int64
Valor                float64
Mantenimiento        float64
Impuesto             float64

Me muestra el numero de filas y columnas del archivo.

In [18]:
dataset_alquileres.shape

(32960, 9)

Obteniendo solo el total de filas.

In [19]:
dataset_alquileres.shape[0]

32960

Obteniendo solo el total de columnas.

In [20]:
dataset_alquileres.shape[1]

9

In [21]:
print("La base de datos presenta {} registros(inmuebles) y {} columnas.".format(dataset_alquileres.shape[0], dataset_alquileres.shape[1]))

La base de datos presenta 32960 registros(inmuebles) y 9 columnas.


Analizando el archivo Alquileres para extraer los valores duplicados si los hay.

In [22]:
dataset_alquileres.head()

Tipo    Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0       Habitación   San Borja        1         0       0    40  1700.0   
1             Casa  Lurigancho        2         0       1   100  7000.0   
2  Local comercial         Ate        0         4       0   150  5200.0   
3     Departamento  Pachacámac        1         0       0    15   800.0   
4     Departamento         Ate        1         0       0    48   800.0   

   Mantenimiento  Impuesto  
0          500.0      60.0  
1            NaN       NaN  
2         4020.0    1111.0  
3          390.0      20.0  
4          230.0       NaN

In [23]:
dataset_alquileres[["Tipo"]]

Tipo
0           Habitación
1                 Casa
2      Local comercial
3         Departamento
4         Departamento
...                ...
32955       Habitación
32956     Departamento
32957     Departamento
32958     Departamento
32959  Local comercial

[32960 rows x 1 columns]

Otra forma de traer columnas para examinar.

In [24]:
dataset_alquileres.Tipo

0             Habitación
1                   Casa
2        Local comercial
3           Departamento
4           Departamento
              ...       
32955         Habitación
32956       Departamento
32957       Departamento
32958       Departamento
32959    Local comercial
Name: Tipo, Length: 32960, dtype: object

In [25]:
tipo_de_inmuebles = dataset_alquileres.Tipo

Es un objeto de tipo series ya que es una sola columna.

In [26]:
type(tipo_de_inmuebles)

pandas.core.series.Series

De esta forma no se eliminan realmente los valores duplicados, solo se estan eliminando para esta visualizacion(debido a que por defecto este metodo contiene inplace = False).

In [27]:
tipo_de_inmuebles.drop_duplicates()

0                Habitación
1                      Casa
2           Local comercial
3              Departamento
7        Casa en condominio
                ...        
4379                Oficina
4721                 Chalet
6983                 Studio
9687                  Hotel
23614      Local industrial
Name: Tipo, Length: 22, dtype: object

Ahora si se eliminaran los duplicados del dataset original.

In [28]:
tipo_de_inmuebles.drop_duplicates(inplace=True)

Se actualizo el contenido original de la variable.

In [29]:
tipo_de_inmuebles

0                Habitación
1                      Casa
2           Local comercial
3              Departamento
7        Casa en condominio
                ...        
4379                Oficina
4721                 Chalet
6983                 Studio
9687                  Hotel
23614      Local industrial
Name: Tipo, Length: 22, dtype: object

#**Organizando la visualizacion del archivo**

In [30]:
tipo_de_inmuebles

0                Habitación
1                      Casa
2           Local comercial
3              Departamento
7        Casa en condominio
                ...        
4379                Oficina
4721                 Chalet
6983                 Studio
9687                  Hotel
23614      Local industrial
Name: Tipo, Length: 22, dtype: object

Visualizando los indices del objeto series.

In [31]:
tipo_de_inmuebles.index

Int64Index([    0,     1,     2,     3,     7,    16,    17,    29,    80,
               83,   117,   159,   207,   347,   589,  2157,  3354,  4379,
             4721,  6983,  9687, 23614],
           dtype='int64')

Lo convierto a df para poder trabajar con los indices, para poder nodificar su comportamiento secuencial.

In [32]:
tipo_de_inmuebles = pd.DataFrame(tipo_de_inmuebles)

In [33]:
tipo_de_inmuebles

Tipo
0              Habitación
1                    Casa
2         Local comercial
3            Departamento
7      Casa en condominio
...                   ...
4379              Oficina
4721               Chalet
6983               Studio
9687                Hotel
23614    Local industrial

[22 rows x 1 columns]

El total de filas de mi df. En otras palabras, se está calculando la longitud del DataFrame en el eje de las filas.

In [34]:
tipo_de_inmuebles.shape[0]

22

Me genera un rango de valores secuenciales(indices) del 0 al 21.

In [35]:
range(tipo_de_inmuebles.shape[0])

range(0, 22)

Una forma mas especifica de visualizar lo mismo, a traves de una iteracion for.

In [36]:
for i in range(tipo_de_inmuebles.shape[0]):
  print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21


**1-- tipo_de_inmuebles:** es un DataFrame que contiene datos sobre tipos de inmuebles, como casas, apartamentos, terrenos, etc.

**2-- tipo_de_inmuebles.shape[0]:** se utiliza para obtener el número de filas en el DataFrame "tipo_de_inmuebles".

**3-- range(tipo_de_inmuebles.shape[0]):** crea un rango de números que va desde 0 hasta el número de filas en el DataFrame "tipo_de_inmuebles". Esto generará una secuencia de números enteros.

Finalmente, se asigna este rango de números como el nuevo índice del DataFrame "tipo_de_inmuebles" utilizando la notación **.index**. Esto significa que en lugar de usar el índice predeterminado basado en las filas originales del DataFrame, ahora se utilizará este nuevo índice numérico.

In [37]:
tipo_de_inmuebles.index = range(tipo_de_inmuebles.shape[0])

In [38]:
tipo_de_inmuebles.index

RangeIndex(start=0, stop=22, step=1)

In [39]:
tipo_de_inmuebles

Tipo
0           Habitación
1                 Casa
2      Local comercial
3         Departamento
4   Casa en condominio
..                 ...
17             Oficina
18              Chalet
19              Studio
20               Hotel
21    Local industrial

[22 rows x 1 columns]

In [40]:
tipo_de_inmuebles.columns.name = "ID"

In [41]:
tipo_de_inmuebles

ID                Tipo
0           Habitación
1                 Casa
2      Local comercial
3         Departamento
4   Casa en condominio
..                 ...
17             Oficina
18              Chalet
19              Studio
20               Hotel
21    Local industrial

[22 rows x 1 columns]

#**Reporte de Analisis II**

**Inmuebles Residenciales**

In [42]:
dataset_alquileres.head()

Tipo    Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0       Habitación   San Borja        1         0       0    40  1700.0   
1             Casa  Lurigancho        2         0       1   100  7000.0   
2  Local comercial         Ate        0         4       0   150  5200.0   
3     Departamento  Pachacámac        1         0       0    15   800.0   
4     Departamento         Ate        1         0       0    48   800.0   

   Mantenimiento  Impuesto  
0          500.0      60.0  
1            NaN       NaN  
2         4020.0    1111.0  
3          390.0      20.0  
4          230.0       NaN

Se excluyen todos los valores duplicados del df pero solo para esta visualizacion.

Obtengo una lista que contiene todos los tipos de propiedades únicos presentes en la columna "Tipo" del DataFrame dataset_alquileres.

In [43]:
list(dataset_alquileres["Tipo"].drop_duplicates())

['Habitación',
 'Casa',
 'Local comercial',
 'Departamento',
 'Casa en condominio',
 'Edificio completo',
 'Flat',
 'Tienda/Salón',
 'Almacén',
 'Casa comercial',
 'Casa de villa',
 'Terreno',
 'Cochera',
 'Loft',
 'Tienda en Centro Comercial',
 'Chacra',
 'Terreno em condominio',
 'Oficina',
 'Chalet',
 'Studio',
 'Hotel',
 'Local industrial']

Creo una lista solo con las propiedades que son residenciales.

In [44]:
residencias = ['Habitación',
 'Casa',
 'Departamento',
 'Casa en condominio',
 'Casa comercial',
 'Casa de villa']
residencias

['Habitación',
 'Casa',
 'Departamento',
 'Casa en condominio',
 'Casa comercial',
 'Casa de villa']

In [45]:
dataset_alquileres.head(10)

Tipo     Distrito  Cuartos  Vacantes  Suites  Area    Valor  \
0          Habitación    San Borja        1         0       0    40   1700.0   
1                Casa   Lurigancho        2         0       1   100   7000.0   
2     Local comercial          Ate        0         4       0   150   5200.0   
3        Departamento   Pachacámac        1         0       0    15    800.0   
4        Departamento          Ate        1         0       0    48    800.0   
5        Departamento     Barranco        3         1       0    70   1200.0   
6        Departamento        Comas        2         0       0    50   1300.0   
7  Casa en condominio          Ate        5         4       5   750  22000.0   
8  Casa en condominio  El Agustino        2         2       0    65   1000.0   
9     Local comercial   Pachacámac        0         3       0   695  35000.0   

   Mantenimiento  Impuesto  
0          500.0      60.0  
1            NaN       NaN  
2         4020.0    1111.0  
3          390.0      20.0  
4          230.0       NaN  
5            NaN       NaN  
6          301.0      17.0  
7            NaN       NaN  
8            NaN       NaN  
9        19193.0    3030.0

**.isin:** va a verificar cada uno de los valores de la columna "Tipo" y ver cuales estan dentro de la lista "residencias", luego devolvera un booleano segun cada caso.

In [46]:
filtro_residencias = dataset_alquileres.Tipo.isin(residencias)

In [47]:
filtro_residencias

0         True
1         True
2        False
3         True
4         True
         ...  
32955     True
32956     True
32957     True
32958     True
32959    False
Name: Tipo, Length: 32960, dtype: bool

Creo un nuevo df que solo contiene registros de inmuebles residenciales.

In [48]:
df_residencias = dataset_alquileres[filtro_residencias]
df_residencias

Tipo     Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0        Habitación    San Borja        1         0       0    40  1700.0   
1              Casa   Lurigancho        2         0       1   100  7000.0   
3      Departamento   Pachacámac        1         0       0    15   800.0   
4      Departamento          Ate        1         0       0    48   800.0   
5      Departamento     Barranco        3         1       0    70  1200.0   
...             ...          ...      ...       ...     ...   ...     ...   
32953  Departamento  Santa Anita        2         0       0    70   900.0   
32955    Habitación   Pachacámac        0         0       0    27   800.0   
32956  Departamento        Lince        3         1       2    78  1800.0   
32957  Departamento        Rímac        2         1       0    48  1400.0   
32958  Departamento        Rímac        2         0       0    70  3000.0   

       Mantenimiento  Impuesto  
0              500.0      60.0  
1                NaN       NaN  
3              390.0      20.0  
4              230.0       NaN  
5                NaN       NaN  
...              ...       ...  
32953          490.0      48.0  
32955          350.0      25.0  
32956          800.0      40.0  
32957          509.0      37.0  
32958          760.0       NaN  

[22845 rows x 9 columns]

Confirmando que se respeto el criterio utilizado para realizar el filtro de los inmuebles.

In [49]:
list(df_residencias["Tipo"].drop_duplicates())

['Habitación',
 'Casa',
 'Departamento',
 'Casa en condominio',
 'Casa comercial',
 'Casa de villa']

In [50]:
df_residencias.shape[0]

22845

Actualizando los indices del df para que vayan en secuencia.

In [51]:
df_residencias.index = range(df_residencias.shape[0])
df_residencias

Tipo     Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0        Habitación    San Borja        1         0       0    40  1700.0   
1              Casa   Lurigancho        2         0       1   100  7000.0   
2      Departamento   Pachacámac        1         0       0    15   800.0   
3      Departamento          Ate        1         0       0    48   800.0   
4      Departamento     Barranco        3         1       0    70  1200.0   
...             ...          ...      ...       ...     ...   ...     ...   
22840  Departamento  Santa Anita        2         0       0    70   900.0   
22841    Habitación   Pachacámac        0         0       0    27   800.0   
22842  Departamento        Lince        3         1       2    78  1800.0   
22843  Departamento        Rímac        2         1       0    48  1400.0   
22844  Departamento        Rímac        2         0       0    70  3000.0   

       Mantenimiento  Impuesto  
0              500.0      60.0  
1                NaN       NaN  
2              390.0      20.0  
3              230.0       NaN  
4                NaN       NaN  
...              ...       ...  
22840          490.0      48.0  
22841          350.0      25.0  
22842          800.0      40.0  
22843          509.0      37.0  
22844          760.0       NaN  

[22845 rows x 9 columns]

#**Exportando la Base de Datos**

In [52]:
df_residencias.to_csv("Alquileres_Residencias.csv", ";")

In [53]:
df_residencias_2 = pd.read_csv("Alquileres_Residencias.csv", sep = ";")
df_residencias_2

Unnamed: 0          Tipo     Distrito  Cuartos  ...  Area   Valor  \
0               0    Habitación    San Borja        1  ...    40  1700.0   
1               1          Casa   Lurigancho        2  ...   100  7000.0   
2               2  Departamento   Pachacámac        1  ...    15   800.0   
3               3  Departamento          Ate        1  ...    48   800.0   
4               4  Departamento     Barranco        3  ...    70  1200.0   
...           ...           ...          ...      ...  ...   ...     ...   
22840       22840  Departamento  Santa Anita        2  ...    70   900.0   
22841       22841    Habitación   Pachacámac        0  ...    27   800.0   
22842       22842  Departamento        Lince        3  ...    78  1800.0   
22843       22843  Departamento        Rímac        2  ...    48  1400.0   
22844       22844  Departamento        Rímac        2  ...    70  3000.0   

       Mantenimiento  Impuesto  
0              500.0      60.0  
1                NaN       NaN  
2              390.0      20.0  
3              230.0       NaN  
4                NaN       NaN  
...              ...       ...  
22840          490.0      48.0  
22841          350.0      25.0  
22842          800.0      40.0  
22843          509.0      37.0  
22844          760.0       NaN  

[22845 rows x 10 columns]

In [54]:
df_residencias_2.head(10)

Unnamed: 0                Tipo     Distrito  Cuartos  ...  Area    Valor  \
0           0          Habitación    San Borja        1  ...    40   1700.0   
1           1                Casa   Lurigancho        2  ...   100   7000.0   
2           2        Departamento   Pachacámac        1  ...    15    800.0   
3           3        Departamento          Ate        1  ...    48    800.0   
4           4        Departamento     Barranco        3  ...    70   1200.0   
5           5        Departamento        Comas        2  ...    50   1300.0   
6           6  Casa en condominio          Ate        5  ...   750  22000.0   
7           7  Casa en condominio  El Agustino        2  ...    65   1000.0   
8           8        Departamento   Pachacámac        1  ...    36   1200.0   
9           9        Departamento      Mi Perú        2  ...    70   1500.0   

   Mantenimiento  Impuesto  
0          500.0      60.0  
1            NaN       NaN  
2          390.0      20.0  
3          230.0       NaN  
4            NaN       NaN  
5          301.0      17.0  
6            NaN       NaN  
7            NaN       NaN  
8            NaN       NaN  
9          642.0      74.0  

[10 rows x 10 columns]

Sobrescribiendo la creacion del metodo de exportacion del archivo alq. residencias para que no me exporte los indices y no se genere esa columna adicional por defecto.

In [55]:
df_residencias.to_csv("Alquileres_Residencias.csv", ";", index=False)

In [56]:
df_residencias_2 = pd.read_csv("Alquileres_Residencias.csv", sep = ";")
df_residencias_2

Tipo     Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0        Habitación    San Borja        1         0       0    40  1700.0   
1              Casa   Lurigancho        2         0       1   100  7000.0   
2      Departamento   Pachacámac        1         0       0    15   800.0   
3      Departamento          Ate        1         0       0    48   800.0   
4      Departamento     Barranco        3         1       0    70  1200.0   
...             ...          ...      ...       ...     ...   ...     ...   
22840  Departamento  Santa Anita        2         0       0    70   900.0   
22841    Habitación   Pachacámac        0         0       0    27   800.0   
22842  Departamento        Lince        3         1       2    78  1800.0   
22843  Departamento        Rímac        2         1       0    48  1400.0   
22844  Departamento        Rímac        2         0       0    70  3000.0   

       Mantenimiento  Impuesto  
0              500.0      60.0  
1                NaN       NaN  
2              390.0      20.0  
3              230.0       NaN  
4                NaN       NaN  
...              ...       ...  
22840          490.0      48.0  
22841          350.0      25.0  
22842          800.0      40.0  
22843          509.0      37.0  
22844          760.0       NaN  

[22845 rows x 9 columns]

#**Reporte de Analisis III**

Selecciones y Frecuencias

In [57]:
df_residencias_2.head(10)

Tipo     Distrito  Cuartos  Vacantes  Suites  Area    Valor  \
0          Habitación    San Borja        1         0       0    40   1700.0   
1                Casa   Lurigancho        2         0       1   100   7000.0   
2        Departamento   Pachacámac        1         0       0    15    800.0   
3        Departamento          Ate        1         0       0    48    800.0   
4        Departamento     Barranco        3         1       0    70   1200.0   
5        Departamento        Comas        2         0       0    50   1300.0   
6  Casa en condominio          Ate        5         4       5   750  22000.0   
7  Casa en condominio  El Agustino        2         2       0    65   1000.0   
8        Departamento   Pachacámac        1         0       0    36   1200.0   
9        Departamento      Mi Perú        2         1       0    70   1500.0   

   Mantenimiento  Impuesto  
0          500.0      60.0  
1            NaN       NaN  
2          390.0      20.0  
3          230.0       NaN  
4            NaN       NaN  
5          301.0      17.0  
6            NaN       NaN  
7            NaN       NaN  
8            NaN       NaN  
9          642.0      74.0

Consulta I

In [58]:
#Seleccione solamente los inmuebles clasificados como "Departamento"

Inspecciono la columna Tipo y comparo(==) todos los registros que tienen como criterio "Departamento".

In [59]:
seleccion_deptos = df_residencias_2["Tipo"] == "Departamento"

In [60]:
seleccion_deptos

0        False
1        False
2         True
3         True
4         True
         ...  
22840     True
22841    False
22842     True
22843     True
22844     True
Name: Tipo, Length: 22845, dtype: bool

De esta forma me quedo solo con los inmuebles que son departamentos.

In [61]:
deptos = df_residencias_2[seleccion_deptos]
deptos

Tipo     Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
2      Departamento   Pachacámac        1         0       0    15   800.0   
3      Departamento          Ate        1         0       0    48   800.0   
4      Departamento     Barranco        3         1       0    70  1200.0   
5      Departamento        Comas        2         0       0    50  1300.0   
8      Departamento   Pachacámac        1         0       0    36  1200.0   
...             ...          ...      ...       ...     ...   ...     ...   
22838  Departamento          Ate        2         0       0    52  1000.0   
22840  Departamento  Santa Anita        2         0       0    70   900.0   
22842  Departamento        Lince        3         1       2    78  1800.0   
22843  Departamento        Rímac        2         1       0    48  1400.0   
22844  Departamento        Rímac        2         0       0    70  3000.0   

       Mantenimiento  Impuesto  
2              390.0      20.0  
3              230.0       NaN  
4                NaN       NaN  
5              301.0      17.0  
8                NaN       NaN  
...              ...       ...  
22838          550.0       NaN  
22840          490.0      48.0  
22842          800.0      40.0  
22843          509.0      37.0  
22844          760.0       NaN  

[19532 rows x 9 columns]

Aunque esta info: **19532 rows × 9 columns** ya nos esta diciendo la frecuencia que tienen los departamento dentro de nuestro df de residencias, lo haremos a continuacion a/t del metodo shape para confirmar.

De esta forma obtenemos la frecuencia que tienen los departamentos dentro de nuestro df_residencias_2.

In [62]:
deptos = df_residencias_2[seleccion_deptos].shape[0]
deptos

19532

Consulta II

In [63]:
#Seleccione los inmuebles clasificados como "Casa", "Casa en condominio" y "Casa de villa"

**seleccion_casas:** es una variable que almacena una condición **booleana**. Esta condición se evalúa como **True** para cada fila en el DataFrame df_residencias_2 si el valor en la columna "Tipo" es igual a "Casa", "Casa en condominio" o "Casa de villa". Es decir, seleccion_casas será True para las filas que representan casas en cualquiera de esas tres categorías y False para las demás filas.

**casas:** es una nueva variable que almacena un DataFrame que contiene únicamente las filas de df_residencias_2 que cumplen con la condición especificada en seleccion_casas. En resumen, casas es un subconjunto de df_residencias_2 que contiene solo las filas que representan casas en alguna categoria de las ya mencionadas.

En este código, se filtran y se seleccionan las casas en base a sus categorías de tipo y se almacenan en la variable casas.






In [64]:
seleccion_casas = (df_residencias_2["Tipo"] == "Casa") | \
 (df_residencias_2["Tipo"] == "Casa en condominio") | \
  (df_residencias_2["Tipo"] == "Casa de villa")
casas = df_residencias_2[seleccion_casas] #"|" => OR
casas

Tipo     Distrito  Cuartos  Vacantes  Suites  Area  \
1                    Casa   Lurigancho        2         0       1   100   
6      Casa en condominio          Ate        5         4       5   750   
7      Casa en condominio  El Agustino        2         2       0    65   
14     Casa en condominio     San Luis        3         1       1   115   
18     Casa en condominio          Ate        4         3       2   466   
...                   ...          ...      ...       ...     ...   ...   
22810  Casa en condominio       Callao        3         3       1   155   
22812                Casa   Carabayllo        2         1       0    60   
22820  Casa en condominio          Ate        3         5       3  1000   
22823                Casa   Pachacámac        5         0       1   160   
22839  Casa en condominio          Ate        5         3       4   450   

         Valor  Mantenimiento  Impuesto  
1       7000.0            NaN       NaN  
6      22000.0            NaN       NaN  
7       1000.0            NaN       NaN  
14      2000.0            NaN       NaN  
18      7500.0         2695.0       NaN  
...        ...            ...       ...  
22810   2000.0          530.0    1250.0  
22812   1400.0            NaN       NaN  
22820  20000.0         3000.0       NaN  
22823   5000.0            NaN      80.0  
22839  15000.0         1711.0    2332.0  

[2212 rows x 9 columns]

In [65]:
casas = df_residencias_2[seleccion_casas].shape[0]
casas

2212

Consulta III

In [66]:
#Seleccione los inmuebles con area entre 60 y 100 metros cuadrados, incluyendo los limites

In [67]:
area_inmuebles = (df_residencias_2["Area"] >= 60) & \
(df_residencias_2["Area"] <= 100)  #"&" => AND
area = df_residencias_2[area_inmuebles]
area   #8736 rows

Tipo     Distrito  Cuartos  Vacantes  Suites  Area  \
1                    Casa   Lurigancho        2         0       1   100   
4            Departamento     Barranco        3         1       0    70   
7      Casa en condominio  El Agustino        2         2       0    65   
9            Departamento      Mi Perú        2         1       0    70   
10           Departamento    San Borja        3         1       1    90   
...                   ...          ...      ...       ...     ...   ...   
22833        Departamento  Jesús María        3         2       1    88   
22835        Departamento        Rímac        3         0       1    80   
22840        Departamento  Santa Anita        2         0       0    70   
22842        Departamento        Lince        3         1       2    78   
22844        Departamento        Rímac        2         0       0    70   

        Valor  Mantenimiento  Impuesto  
1      7000.0            NaN       NaN  
4      1200.0            NaN       NaN  
7      1000.0            NaN       NaN  
9      1500.0          642.0      74.0  
10     1500.0          455.0      14.0  
...       ...            ...       ...  
22833  1550.0          790.0       NaN  
22835  3000.0         1010.0     249.0  
22840   900.0          490.0      48.0  
22842  1800.0          800.0      40.0  
22844  3000.0          760.0       NaN  

[8736 rows x 9 columns]

In [68]:
area = df_residencias_2[area_inmuebles].shape[0]
area

8736

Consulta IV

In [69]:
#Seleccione los que tengan al menos 4 cuartos y alquiler menor a $2.000.00

In [70]:
seleccion_inmuebles = (df_residencias_2["Cuartos"] >= 4) & (df_residencias_2["Valor"] < 2000)
seleccion_1 = df_residencias_2[seleccion_inmuebles]
seleccion_1

Tipo             Distrito  Cuartos  Vacantes  Suites  Area  \
489    Departamento          Jesús María        4         2       2   160   
1020   Departamento             San Luis        4         1       1   110   
1987           Casa                Lince        4         1       4   100   
2129           Casa  Santa María del Mar        4         3       0   120   
2227   Departamento          Jesús María        4         2       2   125   
...             ...                  ...      ...       ...     ...   ...   
20420  Departamento                Lince        4         1       1    80   
21227  Departamento                  Ate        4         3       4   148   
21732  Departamento    Villa El Salvador        4         0       1   180   
22042  Departamento          Jesús María        4         2       2   160   
22198          Casa                Lince        4         2       0   120   

        Valor  Mantenimiento  Impuesto  
489    1900.0          830.0       NaN  
1020   1900.0            NaN       NaN  
1987   1600.0            NaN       NaN  
2129   1500.0            NaN       NaN  
2227   1700.0            NaN      90.0  
...       ...            ...       ...  
20420  1600.0          950.0       NaN  
21227  1500.0         1600.0     196.0  
21732  1900.0            NaN     176.0  
22042  1900.0         1038.0       NaN  
22198  1600.0            NaN       NaN  

[41 rows x 9 columns]

In [71]:
seleccion_1 = df_residencias_2[seleccion_inmuebles].shape[0]
seleccion_1

41

In [72]:
print("Numero de Inmuebles clasificados como 'Departamento' => {}".format(deptos))
print("Numero de Inmuebles clasificados como 'Casa', 'Casa en condominio' y 'Casa de villa' => {}".format(casas))
print("Numero de Inmuebles con area entre 60 y 100 metros cuadrados, incluyendo los limites => {}".format(area))
print("Numero de Inmuebles que tengan por lo menos 4 cuartos y alquiler menor a $2.000 => {}".format(seleccion_1))

Numero de Inmuebles clasificados como 'Departamento' => 19532
Numero de Inmuebles clasificados como 'Casa', 'Casa en condominio' y 'Casa de villa' => 2212
Numero de Inmuebles con area entre 60 y 100 metros cuadrados, incluyendo los limites => 8736
Numero de Inmuebles que tengan por lo menos 4 cuartos y alquiler menor a $2.000 => 41


#**Reporte de Analisis IV**

**Tratamiento de Datos Faltantes I**


In [73]:
dataset_alquileres.head()

Tipo    Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0       Habitación   San Borja        1         0       0    40  1700.0   
1             Casa  Lurigancho        2         0       1   100  7000.0   
2  Local comercial         Ate        0         4       0   150  5200.0   
3     Departamento  Pachacámac        1         0       0    15   800.0   
4     Departamento         Ate        1         0       0    48   800.0   

   Mantenimiento  Impuesto  
0          500.0      60.0  
1            NaN       NaN  
2         4020.0    1111.0  
3          390.0      20.0  
4          230.0       NaN

Devuelve un booleano, en caso de que el campo contenga un valor definido devolvera False, pero si el campo contiene un valor nulo(NaN) devolvera True.

In [74]:
dataset_alquileres.isnull()

Tipo  Distrito  Cuartos  Vacantes  Suites   Area  Valor  \
0      False     False    False     False   False  False  False   
1      False     False    False     False   False  False  False   
2      False     False    False     False   False  False  False   
3      False     False    False     False   False  False  False   
4      False     False    False     False   False  False  False   
...      ...       ...      ...       ...     ...    ...    ...   
32955  False     False    False     False   False  False  False   
32956  False     False    False     False   False  False  False   
32957  False     False    False     False   False  False  False   
32958  False     False    False     False   False  False  False   
32959  False     False    False     False   False  False  False   

       Mantenimiento  Impuesto  
0              False     False  
1               True      True  
2              False     False  
3              False     False  
4              False      True  
...              ...       ...  
32955          False     False  
32956          False     False  
32957          False     False  
32958          False      True  
32959          False     False  

[32960 rows x 9 columns]

Ejecuta la misma logica que el metodo isnull pero al reves.

In [75]:
dataset_alquileres.notnull()

Tipo  Distrito  Cuartos  Vacantes  Suites  Area  Valor  Mantenimiento  \
0      True      True     True      True    True  True   True           True   
1      True      True     True      True    True  True   True          False   
2      True      True     True      True    True  True   True           True   
3      True      True     True      True    True  True   True           True   
4      True      True     True      True    True  True   True           True   
...     ...       ...      ...       ...     ...   ...    ...            ...   
32955  True      True     True      True    True  True   True           True   
32956  True      True     True      True    True  True   True           True   
32957  True      True     True      True    True  True   True           True   
32958  True      True     True      True    True  True   True           True   
32959  True      True     True      True    True  True   True           True   

       Impuesto  
0          True  
1         False  
2          True  
3          True  
4         False  
...         ...  
32955      True  
32956      True  
32957      True  
32958     False  
32959      True  

[32960 rows x 9 columns]

In [76]:
dataset_alquileres.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32960 entries, 0 to 32959
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Tipo           32960 non-null  object 
 1   Distrito       32960 non-null  object 
 2   Cuartos        32960 non-null  int64  
 3   Vacantes       32960 non-null  int64  
 4   Suites         32960 non-null  int64  
 5   Area           32960 non-null  int64  
 6   Valor          32943 non-null  float64
 7   Mantenimiento  28867 non-null  float64
 8   Impuesto       22723 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 2.3+ MB


En este caso este comando no nos ayuda ya que no se aplico ningun filtro sobre la columna valor pero si se genero una condicion.

In [77]:
dataset_alquileres["Valor"].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
32955    False
32956    False
32957    False
32958    False
32959    False
Name: Valor, Length: 32960, dtype: bool

De esta forma me trae todos los registros que en la columna valor figuran valores nulos.

17 en total.

In [78]:
dataset_alquileres[dataset_alquileres["Valor"].isnull()]

Tipo     Distrito  Cuartos  Vacantes  Suites  \
76                   Departamento          Ate        2         1       1   
2186                 Departamento  San Bartolo        2         0       0   
2456                         Casa        Lince        3         4       3   
2646                 Tienda/Salón          Ate        0         2       0   
2941                 Departamento   Chorrillos        2         0       0   
2973                 Departamento    San Borja        2         0       0   
6598                 Tienda/Salón     La Punta        0         0       0   
7302                         Casa        Lince        3         2       1   
12683                Departamento  San Bartolo        2         0       1   
13229                Departamento     La Punta        3         0       1   
13525                Departamento     La Perla        3         1       1   
13579             Local comercial   Chorrillos        0         3       0   
20948             Local comercial       Callao        0         1       0   
21822  Tienda en Centro Comercial          Ate        0         2       0   
23056             Local comercial   Chorrillos        0         0       0   
24805             Local comercial   Chorrillos        0         0       0   
29166             Local comercial   Chorrillos        0         1       0   

       Area  Valor  Mantenimiento  Impuesto  
76       70    NaN          970.0      68.0  
2186     75    NaN          878.0       NaN  
2456    363    NaN            NaN       NaN  
2646     80    NaN            NaN       NaN  
2941     95    NaN         1010.0     170.0  
2973     72    NaN          850.0       NaN  
6598    326    NaN            NaN    1140.0  
7302    100    NaN            NaN       NaN  
12683    75    NaN          878.0       NaN  
13229   135    NaN          910.0     228.0  
13525   105    NaN          880.0     221.0  
13579   200    NaN           10.0       NaN  
20948    24    NaN          500.0      74.0  
21822    77    NaN         3600.0    1390.0  
23056   900    NaN           10.0       NaN  
24805   500    NaN           10.0       NaN  
29166    50    NaN           10.0       NaN

Eliminando todos los registros que contengan valores nulos en la columna valor.

**A = dataset_alquileres.shape[0]:** Se asigna a la variable **A** el número total de filas en el DataFrame dataset_alquileres utilizando el atributo shape y seleccionando el elemento en la posición 0, que representa el número de filas.

**dropna(subset=["Valor"]):** Este método se utiliza para eliminar filas que contienen valores nulos. El argumento subset especifica en qué columnas se deben buscar los valores nulos. En este caso, se está buscando nulos solo en la columna **"Valor"**.

**inplace=True:** Este argumento indica que la modificación debe realizarse directamente en el df original (dataset_alquileres) en lugar de devolver un nuevo DataFrame modificado. Si inplace se establece en True, no es necesario asignar el resultado a una nueva variable; el df original se actualizará directamente.

**B = dataset_alquileres.shape[0]:** Se asigna a la variable **B** el nuevo número total de filas en el DataFrame después de aplicar la condición de filtro.

**A - B:** Se calcula la diferencia entre el número total de filas antes **(A)** y después **(B)** de aplicar el filtro. Esta diferencia representa el número de filas eliminadas por la condición.

En resumen, el código está calculando cuántas filas fueron eliminadas del DataFrame **dataset_alquileres** al aplicar la condición => dataset_alquileres.dropna(subset= ["Valor"], inplace= True)

In [79]:
A = dataset_alquileres.shape[0]
dataset_alquileres.dropna(subset= ["Valor"], inplace= True)
B = dataset_alquileres.shape[0]
A - B

17

Ya no tengo filas con valores nulos en la columna valor.

In [80]:
dataset_alquileres[dataset_alquileres["Valor"].isnull()]

Empty DataFrame
Columns: [Tipo, Distrito, Cuartos, Vacantes, Suites, Area, Valor, Mantenimiento, Impuesto]
Index: []

**Tratamiento de Datos Faltantes II**

In [81]:
dataset_alquileres

Tipo    Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0           Habitación   San Borja        1         0       0    40  1700.0   
1                 Casa  Lurigancho        2         0       1   100  7000.0   
2      Local comercial         Ate        0         4       0   150  5200.0   
3         Departamento  Pachacámac        1         0       0    15   800.0   
4         Departamento         Ate        1         0       0    48   800.0   
...                ...         ...      ...       ...     ...   ...     ...   
32955       Habitación  Pachacámac        0         0       0    27   800.0   
32956     Departamento       Lince        3         1       2    78  1800.0   
32957     Departamento       Rímac        2         1       0    48  1400.0   
32958     Departamento       Rímac        2         0       0    70  3000.0   
32959  Local comercial  Pachacámac        0         0       0   250  6500.0   

       Mantenimiento  Impuesto  
0              500.0      60.0  
1                NaN       NaN  
2             4020.0    1111.0  
3              390.0      20.0  
4              230.0       NaN  
...              ...       ...  
32955          350.0      25.0  
32956          800.0      40.0  
32957          509.0      37.0  
32958          760.0       NaN  
32959         4206.0    1109.0  

[32943 rows x 9 columns]

Filtro todos los registros que tienen valores nulos en la columna mantenimiento.

In [82]:
dataset_alquileres[dataset_alquileres["Mantenimiento"].isnull()]

Tipo     Distrito  Cuartos  Vacantes  Suites  Area  \
1                    Casa   Lurigancho        2         0       1   100   
5            Departamento     Barranco        3         1       0    70   
7      Casa en condominio          Ate        5         4       5   750   
8      Casa en condominio  El Agustino        2         2       0    65   
10           Departamento   Pachacámac        1         0       0    36   
...                   ...          ...      ...       ...     ...   ...   
32914        Tienda/Salón        Lince        0         0       0   160   
32928        Tienda/Salón        Rímac        0         0       0    68   
32929        Tienda/Salón       Callao        0         0       0    80   
32931                Casa   Pachacámac        5         0       1   160   
32954             Cochera   Pachacámac        0         0       0   755   

         Valor  Mantenimiento  Impuesto  
1       7000.0            NaN       NaN  
5       1200.0            NaN       NaN  
7      22000.0            NaN       NaN  
8       1000.0            NaN       NaN  
10      1200.0            NaN       NaN  
...        ...            ...       ...  
32914   2650.0            NaN       NaN  
32928  25000.0            NaN    1113.0  
32929   5800.0            NaN     215.0  
32931   5000.0            NaN      80.0  
32954  14000.0            NaN       NaN  

[4089 rows x 9 columns]

Todo tipo de registros que tienen valor nulo en la columna Mantenimiento-

In [83]:
dataset_alquileres[dataset_alquileres["Mantenimiento"].isnull()].shape[0]

4089

Filtrando por los Registros que presenten valor nulo(NaN) en la columna Mantenimiento y sean de tipo Departamento.

In [84]:
sel_mantenimiento_deptos = (dataset_alquileres["Mantenimiento"].isnull()) & \
(dataset_alquileres["Tipo"]=="Departamento")

**dataset_alquileres = dataset_alquileres[~sel_mantenimiento_deptos]:** Se filtran las filas del DataFrame dataset_alquileres usando la condición **inversa** de sel_mantenimiento_deptos mediante el operador **~.** Esto significa que se está seleccionando las filas que no cumplen con la condición especificada en sel_mantenimiento_deptos.

In [85]:
A = dataset_alquileres.shape[0]
dataset_alquileres = dataset_alquileres[~sel_mantenimiento_deptos]
B = dataset_alquileres.shape[0]
A - B

745

In [86]:
dataset_alquileres[dataset_alquileres["Mantenimiento"].isnull()].shape[0]

3344

**fillna({"Mantenimiento": 0, "Impuesto": 0}):** Este método se utiliza para llenar los valores nulos en el DataFrame. El argumento es un diccionario donde las claves son los nombres de las columnas y los valores son los valores con los que se quiere llenar los nulos. En este caso, estoy llenando los nulos en las columnas "Mantenimiento" e "Impuesto" con el valor 0 en ambos casos.

Después de ejecutar esta línea de código, las filas que originalmente tenían valores nulos en las columnas "Mantenimiento" e "Impuesto" ahora contendrán 0 en lugar de nulos. El DataFrame dataset_alquileres se modificará directamente.

In [87]:
dataset_alquileres = dataset_alquileres.fillna({"Mantenimiento":0, "Impuesto":0})

In [88]:
dataset_alquileres[dataset_alquileres["Mantenimiento"].isnull()].shape[0]

0

In [89]:
dataset_alquileres[dataset_alquileres["Impuesto"].isnull()].shape[0]

0

El df dataset_alquileres ya no tiene valores nulos.

In [90]:
dataset_alquileres.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32198 entries, 0 to 32959
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Tipo           32198 non-null  object 
 1   Distrito       32198 non-null  object 
 2   Cuartos        32198 non-null  int64  
 3   Vacantes       32198 non-null  int64  
 4   Suites         32198 non-null  int64  
 5   Area           32198 non-null  int64  
 6   Valor          32198 non-null  float64
 7   Mantenimiento  32198 non-null  float64
 8   Impuesto       32198 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 2.5+ MB


In [91]:
dataset_alquileres.to_csv("Alquiler_Inmuebles_sin_nulos.csv", sep = ";", index = False)

In [92]:
df_inmuebles_sin_nulos = pd.read_csv("/content/Alquiler_Inmuebles_sin_nulos.csv", sep = ";")
df_inmuebles_sin_nulos

Tipo    Distrito  Cuartos  Vacantes  Suites  Area   Valor  \
0           Habitación   San Borja        1         0       0    40  1700.0   
1                 Casa  Lurigancho        2         0       1   100  7000.0   
2      Local comercial         Ate        0         4       0   150  5200.0   
3         Departamento  Pachacámac        1         0       0    15   800.0   
4         Departamento         Ate        1         0       0    48   800.0   
...                ...         ...      ...       ...     ...   ...     ...   
32193       Habitación  Pachacámac        0         0       0    27   800.0   
32194     Departamento       Lince        3         1       2    78  1800.0   
32195     Departamento       Rímac        2         1       0    48  1400.0   
32196     Departamento       Rímac        2         0       0    70  3000.0   
32197  Local comercial  Pachacámac        0         0       0   250  6500.0   

       Mantenimiento  Impuesto  
0              500.0      60.0  
1                0.0       0.0  
2             4020.0    1111.0  
3              390.0      20.0  
4              230.0       0.0  
...              ...       ...  
32193          350.0      25.0  
32194          800.0      40.0  
32195          509.0      37.0  
32196          760.0       0.0  
32197         4206.0    1109.0  

[32198 rows x 9 columns]